### PR TITLE
chore: .npmrc supply-chain hardening + esbuild 0.19 upgrade

### PR DESCRIPTION
**Pull request by rajathmr2000**

### CODE REVIEW

**Review by Aaryan430 — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -65,7 +65,7 @@
   ],
   "standard-version": {
     "scripts": {
-      "postbump": "npm ci && npm run sri-update && git add doc/rule-descriptions.md"
+      "postbump": "npm ci && npm rebuild esbuild --ignore-scripts=false --foreground-scripts && npm run sri-update && git add doc/rule-descriptions.md"
```
Comment (rajathmr2000) [inline package.json:68]: **[Critical] Release postbump no longer applies the colorjs.io patch**

With `ignore-scripts=true` in the committed `.npmrc`, the `npm ci` here no longer triggers this repo's `prepare` script (`husky && npx patch-package`). `patches/colorjs.io+0.4.3.patch` patches a library bundled into shipped `axe.js` (`lib/core/imports/index.js:8`), so the release-built artifact silently bundles unpatched colorjs.io.

**Suggestion:** insert `npm run prepare` after `npm ci`:

```suggestion
      "postbump": "npm ci && npm run prepare && npm rebuild esbuild --ignore-scripts=false --foreground-scripts && npm run sri-update && git add doc/rule-descriptions.md"
```

The same gap exists in `.github/workflows/update-generated-files.yaml` (bare `npm ci` + `npm run build`, auto-commits output) and — cross-repo — in the umbrella `a11y-engine-core/build/scripts/build_axe.sh`, which needs a coordinated change before the submodule pointer bump.

_Reviewer: stack:code-review_

File: .npmrc [hunk 2]
```diff
@@ -1 +1,11 @@
...
+registry=https://registry.npmjs.org
+
+# Supply-chain hardening directives (AXE-3444). Public repo: access is left
+# unset (access=restricted applies to private repos only).
+# Do NOT add tokens or secrets here; this repo installs from the public registry.
+ignore-scripts=true
+strict-ssl=true
+save-exact=true
+audit-level=high
+engine-strict=true
```
Comment (rajathmr2000) [inline .npmrc:10]: **[Medium] `engine-strict=true` turns dependency engines warnings into hard install failures**

This enforces every dependency's `engines` range at install time. CI pins Node 22 via `.nvmrc`, but the umbrella a11y-engine repo's default toolchain is Node 18.20.4 — a dev running `npm install` inside `axe-core/` on Node 18 may now hard-fail where it previously warned, and `build_axe.sh`'s `nvm use` becomes load-bearing.

**Suggestion:** confirm this is deliberate (developer guide already mandates Node 22+) and call it out in the umbrella repo's setup docs.

_Reviewer: stack:code-review_

File: doc/developer-guide.md [hunk 3]
```diff
@@ -34,6 +34,7 @@ Axe 3.0 supports open Shadow DOM: see our virtual DOM APIs and test utilities fo
 1. You must have Node.js version 22 or higher installed.
    If you have [nvm](https://github.com/nvm-sh/nvm) installed, simply do `nvm use` in the root of this repository.
 1. Install npm development dependencies. In the root folder of your axe-core repository, run `npm install`
+1. The committed `.npmrc` sets `ignore-scripts=true` for supply-chain hardening, so dependency lifecycle scripts do not run during install. After installing, run `npm run prepare` (applies dependency patches and installs git hooks) and `npm rebuild esbuild --ignore-scripts=false` (fetches the esbuild bundler binary) before building.
```
Comment (rajathmr2000) [inline doc/developer-guide.md:37]: **[Low] Slightly inaccurate description of the esbuild rebuild**

With esbuild ≥0.16 the platform binary arrives as an `@esbuild/*` optionalDependency during `npm ci`; `npm rebuild esbuild` validates/links it rather than "fetches" it. The command here also omits `--foreground-scripts` used in every CI invocation.

**Suggestion:** reword to "validates and links the esbuild platform binary" and use the full flag set for consistency.

_Reviewer: stack:code-review_

**Review by rajathmr2000 — COMMENTED** [review]
Claude Code Review (automated) — 3 inline finding(s). Full report in the PR comment below. Verdict: Failed - see PR comment.

**Comment by rajathmr2000** [comment]
# Claude Code PR Review

**PR:** https://github.com/browserstack/a11y-engine-axe-core/pull/238  •  **Head:** bc587ba  •  **Reviewers:** stack:code-review

## Summary
Hardens the committed `.npmrc` with supply-chain directives (`ignore-scripts=true`, `save-exact=true`, `engine-strict=true`, etc.), upgrades esbuild 0.10.2 → 0.19.12 (exact-pinned), and adds `npm rebuild esbuild --ignore-scripts=false` follow-ups to the patched CI install paths (GH `test.yml` build job, CircleCI dependencies job) plus the standard-version `postbump` chain; documents the new local-dev steps in the developer guide.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | `.npmrc` carries no tokens and explicitly documents that none may be added |
| High | Security | Authentication/authorization checks present | N/A | Config/CI/docs-only change; no routes or sockets |
| High | Security | Input validation and sanitization | N/A | No runtime code touched |
| High | Security | No IDOR — resource ownership validated | N/A | — |
| High | Security | No SQL injection (parameterized queries) | N/A | — |
| High | Correctness | Logic is correct, handles edge cases | Fail | `ignore-scripts=true` suppresses the repo's own `prepare` script (`husky && npx patch-package`); `patches/colorjs.io+0.4.3.patch` patches a library bundled into shipped `axe.js`. The release `postbump` chain, `update-generated-files.yaml`, and the umbrella repo's `build_axe.sh` all lost the patch silently |
| High | Correctness | Error handling is explicit, no swallowed exceptions | N/A | No runtime code; the failure mode here is precisely that nothing errors — builds succeed with unpatched colorjs.io |
| High | Correctness | No race conditions or concurrency issues | N/A | — |
| Medium | Testing | New code has corresponding tests | N/A | CI/config change; validated by CI itself |
| Medium | Testing | Error paths and edge cases tested | Fail | No release dry-run evidence (`standard-version --dry-run` / `next-release`) proving the postbump chain produces a correctly patched artifact under `ignore-scripts=true` |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | All 12 checks green on head `bc587ba` (build, fmt_check, test_node 6/18/20/22, CodeQL, semgrep, SCA) |
| Medium | Performance | No N+1 queries or unbounded data fetching | N/A | — |
| Medium | Performance | Long-running tasks use background jobs | N/A | — |
| Medium | Quality | Follows existing codebase patterns | Pass | esbuild 0.19 invocation verified compatible (`build/tasks/esbuild.js` uses only API-stable options); lockfile v1+v2 sections updated consistently; exact pin matches `save-exact=true` |
| Medium | Quality | Changes are focused (single concern) | Pass | One concern: install-time script hardening + the esbuild bump it forces |
| Low | Quality | Meaningful names, no dead code | Pass | — |
| Low | Quality | Comments explain why, not what | Pass | `.npmrc` comments explain rationale (public-repo `access` caveat, no-tokens rule) |
| Low | Quality | No unnecessary dependencies added | Pass | New `@esbuild/*` lockfile entries are esbuild 0.19's own platform optionalDependencies |

## Findings

- **File:** `package.json:68`
- **Severity:** Critical
- **Reviewer:** stack:code-review
- **Issue:** The standard-version `postbump` runs `npm ci && npm rebuild esbuild … && npm run sri-update`, but with `ignore-scripts=true` the `npm ci` no longer triggers the repo's `prepare` script (`husky && npx patch-package`). `patches/colorjs.io+0.4.3.patch` patches colorjs.io, which is bundled into `axe.js` (`lib/core/imports/index.js:8`) — so the release-built artifact silently bundles **unpatched** colorjs.io.
- **Suggestion:** Insert `npm run prepare` (or at minimum `npm run patch`) after `npm ci` in `postbump`: `npm ci && npm run prepare && npm rebuild esbuild --ignore-scripts=false --foreground-scripts && npm run sri-update && git add doc/rule-descriptions.md`.

- **File:** `.github/workflows/update-generated-files.yaml:26`
- **Severity:** High
- **Reviewer:** stack:code-review
- **Issue:** This workflow runs `npm ci` + `npm run build` and auto-commits generated files, but was not given the `npm rebuild esbuild` + `npm run prepare` follow-ups added to `test.yml` — same unpatched-colorjs.io problem, with the output committed back to the repo. (File not in this PR's diff.)
- **Suggestion:** Add the same two steps after `npm ci` as in `test.yml:21-22`.

- **File:** `a11y-engine-core/build/scripts/build_axe.sh` (umbrella repo — cross-repo)
- **Severity:** High
- **Reviewer:** stack:code-review
- **Issue:** The umbrella repo's engine build does `npm install` + `npm run build` inside the `axe-core/` submodule with no `prepare`/rebuild step. Once this `.npmrc` lands in the submodule, every a11y-engine-core build bundles unpatched colorjs.io (loses the `CSS?.supports` guard affecting color-contrast paths). Cannot be fixed in this PR.
- **Suggestion:** Sequence a coordinated AXE-3443/AXE-3444 change to `build_axe.sh` (add `npm run prepare && npm rebuild esbuild --ignore-scripts=false`) **before** bumping the submodule pointer, and state the dependency explicitly in the PR description.

- **File:** `.github/workflows/release.yml:26`
- **Severity:** Medium
- **Reviewer:** stack:code-review
- **Issue:** The release job's correctness depends entirely on the `postbump` chain in `package.json` for prepare/rebuild. (File not in this PR's diff.)
- **Suggestion:** After fixing `postbump`, also add an explicit `npm run prepare` to the release workflow so artifact correctness doesn't hinge on standard-version config.

- **File:** `.npmrc:10`
- **Severity:** Medium
- **Reviewer:** stack:code-review
- **Issue:** `engine-strict=true` enforces every dependency's `engines` range at install time, turning previous warnings into hard install failures. CI pins Node 22 via `.nvmrc`, but the umbrella repo's default toolchain is Node 18.20.4 — a dev running `npm install` inside `axe-core/` on Node 18 may now hard-fail, and `build_axe.sh`'s `nvm use` becomes load-bearing.
- **Suggestion:** Confirm this is deliberate (the developer guide already mandates Node 22+); if so, note it in the umbrella repo's setup docs.

- **File:** `.circleci/config.yml:207,223`
- **Severity:** Medium
- **Reviewer:** stack:code-review
- **Issue:** `npm install w3c/wcag-act-rules#main` / `w3c/aria-practices#main` are git dependencies whose own `prepare` scripts are skipped under `ignore-scripts=true`. Both look content-only today, but the `test:act`/`test:apg` jobs were given no follow-up step. (Lines not in this PR's diff.)
- **Suggestion:** Confirm both CircleCI jobs pass on this branch; add a targeted `--ignore-scripts=false` to those installs if they break.

- **File:** `doc/developer-guide.md:37`
- **Severity:** Low
- **Reviewer:** stack:code-review
- **Issue:** "fetches the esbuild bundler binary" is slightly inaccurate — with esbuild ≥0.16 the platform binary arrives as an `@esbuild/*` optionalDependency during `npm ci`; the rebuild validates/links it. The documented command also omits `--foreground-scripts` used everywhere else.
- **Suggestion:** Reword to "validates and links the esbuild platform binary" and include the full flag set for consistency.

Positive notes from the reviewer: the esbuild 0.10 → 0.19 jump itself is safe (the sole invocation in `build/tasks/esbuild.js` uses only options stable across the 0.17 breaking changes); the CircleCI rebuild line is correctly placed before `save_cache` so downstream jobs inherit the rebuilt binary; CircleCI's build job already runs `npm run prepare` explicitly, so the patch-package gap does not apply there; submodule `// [tag]:` rule treated as satisfied (no JS source changed, no sensible anchor for inline tags, no sensitive info in `.npmrc` comments, no Deque packages).

Reviewer questions for the PR thread: (1) Has a full `npm ci && npm run prepare && npm run build && npm test` been run from a clean clone on this branch? (2) Was a release dry-run executed to prove the postbump chain end-to-end under `ignore-scripts`? (3) Is the umbrella `build_axe.sh` change merged/sequenced before the submodule pointer bump?

---

**Verdict: FAIL** — the release postbump chain and two other install→build paths silently lose the bundled colorjs.io patch under `ignore-scripts=true`; fix `postbump` and `update-generated-files.yaml` in this PR and sequence the umbrella `build_axe.sh` change before the pointer bump.

File: package.json (hunk 1 above)
Comment (rajathmr2000) [reply package.json:68]: Tested there was no impact with this

File: doc/developer-guide.md (hunk 3 above)
Comment (rajathmr2000) [reply doc/developer-guide.md:37]: This is fine
